### PR TITLE
Upgrade react-router-dom from v5.3.0 to v6.22.3

### PR DESCRIPTION
**Pull request by bau123**
## Description of change

Update datahub FE react-router-dom from v5 to v6

Main changes include:

- React router dom. v6 make use of hooks instead of wrapping components in a Route component
- No longer using `history` instead use `navigate` from import `react-router-dom`
- Change `<Redirect/>` component to` <Navigate/>`
- Swap `props.params` to `UseParams`
- No longer need to use `<Route />` component to access `location`. Import `location` from `react-router-dom` instead
- Using `<Routes />` instead of `<Switch />` _Note: Children of Routes can only be of `<Route>` component_
- swap `connected-react-router `for `redux-first-history`
- `Routes.js` now takes in element instead of component
- Removed `withRouter` as it is deprecated
- No need to create DataHubProvider in component test. Cypress can now manage this with `cy.mountWithProvider(<Component/>)`

Read more about the [upgrade v5 to v6]( https://reactrouter.com/en/main/upgrading/v5)
Read more about swapping [connected-react-router for redux-first-history](https://github.com/uktrade/data-hub-frontend/pull/4971)
## Checklist

[//]: # 'When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/main/docs/Code%20review%20guidelines.md'

- [ ] Has the branch been rebased to main?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or End-to-End)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, Edge, Safari)

### CODE REVIEW

File: src/client/modules/ExportPipeline/ExportForm/withRouter.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import { useLocation, useNavigate, useParams } from 'react-router-dom'
+
+function withRouter(Component) {
+  function ComponentWithRouterProp(props) {
+    let location = useLocation()
+    let navigate = useNavigate()
+    let params = useParams()
```
Comment (paulgain) [inline src/client/modules/ExportPipeline/ExportForm/withRouter.js:8]: For consistency with the rest of the app these should be `const`.

File: src/client/utils/multiinstance.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 import React from 'react'
-import { withRouter } from 'react-router-dom'
+import {} from 'react-router-dom'
```
Comment (paulgain) [inline src/client/utils/multiinstance.js:2]: Looks like we can remove this line.

**Comment by paulgain** [comment]
Looks good @bau123, just a couple of minor comments. Considering the size of this PR feel free to create a separate PR for these changes or update in this PR - your call.

**Review by paulgain — APPROVED** [review]

**Review by peterhudec — APPROVED** [review]
Looks great, you've done an amazing job! 🥇 I lef a couple of comments, but as @paulgain said, feel free to address them in separate PRs.

File: src/client/components/NotificationAlert/reducer.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { REMINDER_SUMMARY__LOADED } from '../../actions'
 
-const initialState = {
+export const INITIAL_STATE = {
```
[the remark this answers is not in the document]
Comment (paulgain) [reply src/client/components/NotificationAlert/reducer.js:3]: We changed this @peterhudec, we import it into the `mountWithProvider` command found [here](https://github.com/uktrade/data-hub-frontend/blob/react-router-dom-upgrade/test/component/cypress/support/commands.js#L5) :smile:

File: src/client/modules/ExportPipeline/ExportForm/withRouter.js (hunk 1 above)
Comment (bau123) [reply src/client/modules/ExportPipeline/ExportForm/withRouter.js:8]: Will create a new PR for suggested. hanges

File: docs/Data Fetching Building Blocks.md [hunk 4]
```diff
@@ -25,36 +25,40 @@ and delegate the call to the `Task` component. To what _task_ function exactly
...
+<DataHubProvider
+  tasks={{
+    // Satisfy the dependecies
+    'do something': (payload) => Promise.resolve('result'),
+    'do something else': (payload) => Promise.reject('Error message'),
+    // 'talk to the API': () => apiProxyAxios.get('/some-endpoint'),
+    'talk to the API': () =>
+      new Promise((resolve, reject) =>
+        setTimeout(5000, resolve, { dummy: 'data' })
+      ),
```
[the remark this answers is not in the document]
Comment (bau123) [reply docs/Data Fetching Building Blocks.md:44]: Will create a new PR for suggested changes

File: docs/Data Fetching Building Blocks.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Data Fetching Building Blocks
 
-**TLDR** 
+**TLDR**
```
[the remark this answers is not in the document]
Comment (bau123) [reply docs/Data Fetching Building Blocks.md:3]: Will create a new PR for suggested changes

File: src/client/components/NotificationAlert/reducer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (bau123) [reply src/client/components/NotificationAlert/reducer.js:3]: Will create a new PR for suggested changes

File: src/client/components/ProtectedRoute/index.jsx [hunk 6]
```diff
@@ -11,13 +10,14 @@ export const ProtectedRoute = ({
...
   routePermissions = [],
   ...rest
-}) =>
-  routePermissions.every((p) => userPermissions.includes(p)) &
-  modulePermissions.includes(module) ? (
-    <Route {...rest} />
+}) => {
+  return routePermissions.every((p) => userPermissions.includes(p)) &
+    modulePermissions.includes(module) ? (
+    <>{rest.children}</>
```
[the remark this answers is not in the document]
Comment (bau123) [reply src/client/components/ProtectedRoute/index.jsx:16]: Will create a new PR for suggested changes

File: src/client/modules/ExportPipeline/ExportForm/__test__/state.test.js [hunk 7]
```diff
@@ -155,7 +155,7 @@ describe('state2props', () => {
       expect(
         state2props(
           { [COMPANY_DETAILS_ID]: {}, [EXPORT_DETAILS_ID]: {} },
-          { location: {} }
+          { router: { location: {} } }
```
[the remark this answers is not in the document]
Comment (bau123) [reply src/client/modules/ExportPipeline/ExportForm/__test__/state.test.js:158]: Will create a new PR for suggested changes

File: src/client/modules/ExportPipeline/ExportForm/withRouter.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    return <Component {...props} router={{ location, navigate, params }} />
+  }
```
[the remark this answers is not in the document]
Comment (bau123) [reply src/client/modules/ExportPipeline/ExportForm/withRouter.js:10]: Will create a new PR for suggested changes

File: src/middleware/api-proxy.js [hunk 9]
```diff
@@ -57,6 +57,7 @@ const ALLOWLIST = [
   '/v3/event/',
   '/v4/event/',
   '/v4/export-win/',
+  '/v4/export-win/review/:token',
```
[the remark this answers is not in the document]
Comment (bau123) [reply src/middleware/api-proxy.js:60]: Will create a new PR for suggested changes

**Pull request merged**